pkg/util/cloudfoundry/types_test.go

```diff
@@ -0,0 +1,181 @@
+	CellID:      "cell123",
+	ContainerIP: "1.2.3.5",
+	Index:       3,
+	Ports:       []uint32{1234, 5678},
+	ProcessGUID: "0123456789012345678901234567890123456789",
+	State:       "RUNNING",
+}
+
+var BBSModelD1 = models.DesiredLRP{
+	ProcessGuid: "0123456789012345678901234567890123456789",
+	Action: &models.Action{
+		CodependentAction: &models.CodependentAction{
+			Actions: []*models.Action{
+				{
+					RunAction: &models.RunAction{
+						// this run action will be ignored, as it only has VCAP_SERVICES
+						Env: []*models.EnvironmentVariable{
+							{
+								Name:  "VCAP_SERVICES",
+								Value: "{\"broker\": [{\"name\": \"yyy\"}]}",
+							},
+						},
+					},
+				},
+				{
+					RunAction: &models.RunAction{
+						Env: []*models.EnvironmentVariable{
+							{
+								Name:  "we ignore this",
+								Value: "some value",
+							},
+							{
+								Name:  "AD_DATADOGHQ_COM",
+								Value: "{\"xxx\": {}}", // make this a valid JSON
+							},
+							{
+								Name:  "VCAP_SERVICES",
+								Value: "{\"broker\": [{\"name\": \"xxx\"}]}",
+							},
+						},
+					},
+				},
+			},
+		},
+	},
+}
+
+var ExpectedD1 = DesiredLRP{
+	AppGUID:         "012345678901234567890123456789012345",
+	EnvAD:           ADConfig{"xxx": {}},
+	EnvVcapServices: map[string][]byte{"xxx": []byte("{\"name\":\"xxx\"}")},
+	ProcessGUID:     "0123456789012345678901234567890123456789",
+}
+
+func TestADIdentifier(t *testing.T) {
+	for _, tc := range []struct {
+		dLRP     DesiredLRP
+		svcName  string
+		aLRP     *ActualLRP
+		expected string
+	}{
+		{
+			dLRP: DesiredLRP{
+				ProcessGUID: "1234",
+			},
+			svcName:  "postgres",
+			aLRP:     nil,
+			expected: "1234/postgres",
+		},
+		{
+			dLRP: DesiredLRP{
+				ProcessGUID: "1234",
+			},
+			svcName: "flask-app",
+			aLRP: &ActualLRP{
+				Index: 2,
+			},
+			expected: "1234/flask-app/2",
+		},
+	} {
+		t.Run(fmt.Sprintf(""), func(t *testing.T) {
+			var i ADIdentifier
+			if tc.aLRP == nil {
+				i = NewADNonContainerIdentifier(tc.dLRP, tc.svcName)
+			} else {
+				i = NewADContainerIdentifier(tc.dLRP, tc.svcName, *tc.aLRP)
+			}
+			assert.EqualValues(t, tc.expected, i.String())
+		})
+	}
+}
+
+func TestActualLRPFromBBSModel(t *testing.T) {
+	result := ActualLRPFromBBSModel(&BBSModelA1)
+	assert.EqualValues(t, ExpectedA1, result)
+}
+
+func TestDesiredLRPFromBBSModel(t *testing.T) {
+	result := DesiredLRPFromBBSModel(&BBSModelD1)
+	assert.EqualValues(t, ExpectedD1, result)
+}
+
+func TestGetVcapServicesMap(t *testing.T) {
+	input := `{
+		"broker1": [{"name": "s1", "attr1": "a1"}, {"name": "s2", "attr2": "a2"}],
+		"broker2": [{"name": "s3", "attr3": "a3"}]
+	}`
+	expected := map[string][]byte{
+		"s1": []byte(`{"attr1":"a1","name":"s1"}`),
+		"s2": []byte(`{"attr2":"a2","name":"s2"}`),
+		"s3": []byte(`{"attr3":"a3","name":"s3"}`),
+	}
+	result, err := getVcapServicesMap(input, "xxx")
+	assert.Nil(t, err)
+	assert.EqualValues(t, expected, result)
+}
```

Comment:
Should we test the String method as well. This comment applies everywhere here as well. We should test all methods even if they are simple.

Reply:
Good point, I added that method later and forgot to add tests for it.

Reply:
Ah wait, it *is* tested, see line 153.